server/src/channels/commands/clean_personal_access_tokens.rs

```diff
@@ -0,0 +1,144 @@
+        config: &PersonalAccessTokenCleanerConfig,
+        sender: Sender<CleanPersonalAccessTokensCommand>,
+    ) -> Self {
+        Self {
+            enabled: config.enabled,
+            interval: Duration::from_secs(config.interval),
+            sender,
+        }
+    }
+
+    pub fn start(&self) {
+        if !self.enabled {
+            info!("Personal access token cleaner is disabled.");
+            return;
+        }
+
+        let interval = self.interval;
+        let sender = self.sender.clone();
+        info!(
+            "Personal access token cleaner is enabled, expired tokens will be deleted every: {:?}.",
+            interval
+        );
+
+        tokio::spawn(async move {
+            let mut interval_timer = time::interval(interval);
+            loop {
+                interval_timer.tick().await;
+                if sender.send(CleanPersonalAccessTokensCommand).is_err() {
+                    error!("Failed to send CleanPersonalAccessTokensCommand");
+                }
+            }
+        });
+    }
+}
+
+#[async_trait]
+impl ServerCommand<CleanPersonalAccessTokensCommand> for CleanPersonalAccessTokensExecutor {
+    async fn execute(
+        &mut self,
+        system: &Arc<RwLock<System>>,
+        _command: CleanPersonalAccessTokensCommand,
+    ) {
+        let system = system.read().await;
+        let tokens = system.storage.user.load_all_pats().await;
+        if tokens.is_err() {
+            error!("Failed to load personal access tokens: {:?}", tokens);
+            return;
+        }
+
+        let tokens = tokens.unwrap();
+        if tokens.is_empty() {
+            debug!("No personal access tokens to delete.");
+            return;
+        }
+
+        let now = TimeStamp::now().to_micros();
+        let expired_tokens = tokens
+            .into_iter()
+            .filter(|token| token.is_expired(now))
+            .collect::<Vec<_>>();
+
+        if expired_tokens.is_empty() {
+            debug!("No expired personal access tokens to delete.");
+            return;
+        }
+
+        let expired_tokens_count = expired_tokens.len();
+        debug!("Found {expired_tokens_count} expired personal access tokens.");
+        for token in expired_tokens {
+            let result = system
+                .storage
+                .user
+                .delete_pat(token.user_id, &token.name)
+                .await;
+            if result.is_err() {
+                error!(
+                    "Failed to delete personal access token: {} for user with ID: {}. Error: {:?}",
+                    token.name,
+                    token.user_id,
+                    result.err().unwrap()
+                );
+                continue;
+            }
+
+            debug!(
+                "Deleted personal access token: {} for user with ID: {}.",
+                token.name, token.user_id
+            );
+        }
+
+        info!("Deleted {expired_tokens_count} expired personal access tokens.");
```

Comment:
In case of token deletion failure this will not be true - maybe count how many has been deleted and print deleted/total_to_delete? Not a critical - just a thought.

Reply:
Valid point, fixed :)